ocaml/typing/types.mli

```diff
@@ -673,6 +673,7 @@ module type Wrapped = sig
   | Mty_signature of signature
   | Mty_functor of functor_parameter * module_type
   | Mty_alias of Path.t
+  | Mty_strengthen of module_type * Path.t * bool
```

Comment:
None of the syntax tree sum types are records so this would really go against the style in this module. Honestly, I'd rather not do that. Also, in Haskell it is quite unusual to have a sum type where some constructors are records and some aren't. Isn't it the same in OCaml? I'll add a comment as to what it does. I'm not really sure this is the right place to explain aliasable vs. non-aliasable strengthening, though. This isn't a new concept, this just moves an already existing flag into the syntax tree.
It's not an optimisation and you can't calculate it from the other arguments. It is stable under substitution.

Reply:
As for the type: Adding an unlabeled bool to the type algebra seems like a clear wart.  It looks to me like there are a number of records in this module, and also in `typed_tree`.  But if you don't want a record because of the lack of uniformity, another solution would be to make this a variant instead of a bool.

As for the meaning: Thanks, I think I'm finally starting to understand.  When I read this code initially, to figure out the meaning of this bool I went looking for places where it is determined.  The places that looked most relevant were `Typemod.approx_modtype_jane_syntax` and `Typemod.transl_modtype_jane_syntax_aux`, where this is computed based on the path.  But I think you are saying the more interesting cases are the ones in `Mtype.strengthen_*`, where we're recording information that can't otherwise be determined.  I will re-read with this in mind.

Reply:
Re. OCaml vs. Haskell: A variant with one constructor taking a record is weird in Haskell because it generates top-level partial record projection functions, which you ~never want.  In OCaml, you can't project out without pattern-matching first -- it's equivalent to creating a fresh record type and having a single-argument constructor that takes it, but you can't let that record type escape.  (And there's one less pointer.)

Reply:
I added a separate type which represents aliasability. I also added a comment in mtype.ml which explains when strengthening is aliasable. Do you think that's enough?

@antalsz Sure, but I'd still find this weird:
```
type t =
  | T of int * int * int
  | U of { foo : int; bar : int }
```